From 9ae4d23cb2e2f8aad83073d0d935ff2f3a5600cf Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Thu, 27 Aug 2026 12:26:13 +0100
Subject: [PATCH] input/input_nfc.ml: Increase wait for nbdkit timeout to 5
 minutes

This is a workaround until we can fix the plugin.

Fixes: https://redhat.atlassian.net/browse/RHEL-249754
(cherry picked from commit 31fdc159047f0e0b803b2da4bea3ca6917adc02f)
---
 input/input_nfc.ml | 15 ++++++++++-----
 lib/nbdkit.ml      |  5 ++++-
 lib/nbdkit.mli     |  4 ++++
 3 files changed, 18 insertions(+), 6 deletions(-)

diff --git a/input/input_nfc.ml b/input/input_nfc.ml
index 8ee4b606..de70d63a 100644
--- a/input/input_nfc.ml
+++ b/input/input_nfc.ml
@@ -154,11 +154,10 @@ these settings.
       match options.input_password with
       | None ->
          (* Because we will start nbdkit in the background and then wait
-          * for 30 seconds for it to start up, we cannot use the
-          * password=- feature of nbdkit to read the password
-          * interactively (since in the words of the movie the user has
-          * only "30 seconds to comply").  So in the
-          * AskForPassword case we read the password here.
+          * for it to start up, we cannot use the password=- feature of
+          * nbdkit to read the password interactively (since in the words
+          * of the movie the user has only "30 seconds to comply").  So
+          * in the AskForPassword case we read the password here.
           *)
          printf (f_"%s: enter password for ‘%s’: ") "nfc" user;
          let open Unix in
@@ -231,6 +230,12 @@ See also the virt-v2v-input-vmware(1) manual.")
     let create_nbdkit_nfc ?name () =
       let cmd = Nbdkit.create ?name "nfc" in
 
+      (* Work around the long time that nbdkit-nfc-plugin takes to
+       * start up when the server is distant.  Plan to fix this properly
+       * in the plugin sometime (RHEL-249752).
+       *)
+      Nbdkit.set_timeout cmd (5 * 60);
+
       Nbdkit.add_arg cmd "server" server;
       (* I'm pretty sure this was the default with VDDK.  Maybe it
        * should be settable via -io?
diff --git a/lib/nbdkit.ml b/lib/nbdkit.ml
index e975bddf..3e8c077f 100644
--- a/lib/nbdkit.ml
+++ b/lib/nbdkit.ml
@@ -105,6 +105,7 @@ type cmd = {
   mutable env : (string * string) list;
   mutable debug_flags : (string * string) list;
   mutable readonly : bool;
+  mutable timeout : int; (* seconds *)
   mutable threads : int;
   verbose : bool;
 }
@@ -118,6 +119,7 @@ let create ?(quiet = false) ?name plugin = {
   debug_flags = [];
   readonly = false;
   threads = 16;
+  timeout = 30;
   verbose = not quiet && verbose ()
 }
 
@@ -126,6 +128,7 @@ let add_debug_flag cmd name value =
 
 let set_readonly cmd v = cmd.readonly <- v
 let set_threads cmd v = cmd.threads <- v
+let set_timeout cmd v = cmd.timeout <- v
 let add_filter cmd v = cmd.filters <- v :: cmd.filters
 let add_arg cmd key value = cmd.args <- (key, value) :: cmd.args
 let add_args cmd kvs = cmd.args <- List.rev kvs @ cmd.args
@@ -218,7 +221,7 @@ let run_unix socket cmd =
   (* Wait for the pidfile to appear so we know that nbdkit
    * is listening for requests.
    *)
-  if not (wait_for_file pidfile 30) then (
+  if not (wait_for_file pidfile cmd.timeout) then (
     if verbose () then
       error (f_"nbdkit did not start up.  See previous debugging messages for problems.")
     else
diff --git a/lib/nbdkit.mli b/lib/nbdkit.mli
index eaf65f05..bfd90107 100644
--- a/lib/nbdkit.mli
+++ b/lib/nbdkit.mli
@@ -94,6 +94,10 @@ val set_threads : cmd -> int -> unit
 (** Set the number of threads ([--threads] option).
     The default is 16. *)
 
+val set_timeout : cmd -> int -> unit
+(** Set the number of seconds to wait for nbdkit to start up.
+    The default is 30 seconds. *)
+
 val add_filter : cmd -> string -> unit
 (** Add a filter.  You may need to use {!probe_filter} first to check
     the filter is installed, otherwise nbdkit will fail to run.
